File: src/main/java/org/example/internals/math/Matrix.java
package org.example.internals.math;

/**
 * Represents a two-dimensional integer matrix and provides basic
 * linear algebra operations.
 * This matrix implementation supports addition, multiplication,
 * transposition, inversion (via Gaussian elimination),
 * and determinant calculation.
 * All values are stored as integers. Some operations (e.g. inversion)
 * may lose precision due to integer division.
 */
public class Matrix {

  private final int[][] numbers;
  private final int rows;
  private final int columns;

  /**
   * Creates a new matrix with the given dimensions.
   *
   * @param rows    number of rows
   * @param columns number of columns
   */
  public Matrix(int rows, int columns) {
    this.rows = rows;
    this.columns = columns;
    this.numbers = new int[rows][columns];
  }

  private Matrix(int[][] numbers) {
    this.numbers = numbers;
    this.rows = numbers.length;
    this.columns = numbers[0].length;
    for (int i = 0; i < rows; i++) {
      System.arraycopy(numbers[i], 0, this.numbers[i], 0, columns);
    }
  }

  /**
   * Returns the number of columns.
   *
   * @return column count
   */
  public int getColumns() {
    return columns;
  }

  /**
   * Returns the number of rows.
   *
   * @return row count
   */
  public int getRows() {
    return rows;
  }

  /**
   * Sets the value at the given matrix position.
   *
   * @param i     row index
   * @param j     column index
   * @param value value to set
   */
  public void setNumber(int i, int j, int value) {
    numbers[i][j] = value;
  }

  /**
   * Returns the value at the given matrix position.
   *
   * @param i row index
   * @param j column index
   * @return value at the specified position
   */
  public int getNumber(int i, int j) {
    return numbers[i][j];
  }

  /**
   * Adds this matrix to another matrix.
   *
   * @param other matrix to add
   * @return resulting matrix
   * @throws IllegalArgumentException if matrix dimensions do not match
   */
  public Matrix add(Matrix other) {
    if (this.rows == other.rows && this.columns == other.columns) {
      Matrix ret = new Matrix(rows, columns);
      for (int i = 0; i < rows; i++) {
        for (int j = 0; j < columns; j++) {
          ret.setNumber(i, j, this.getNumber(i, j) + other.getNumber(i, j));
        }
      }
      return ret;
    }
    throw new IllegalArgumentException("Unable to perform add");
  }

  /**
   * Multiplies this matrix with another matrix.
   *
   * @param other matrix to multiply with
   * @return resulting matrix
   * @throws IllegalArgumentException if matrix dimensions are incompatible
   */
  public Matrix multiply(Matrix other) {
    if (this.rows == other.columns && this.columns == other.rows) {
      Matrix ret = new Matrix(this.rows, other.columns);
      for (int i = 0; i < ret.rows; i++) {
        for (int j = 0; j < ret.columns; j++) {
          int sum = 0;
          for (int k = 0; k < this.columns; k++) {
            sum += this.getNumber(i, k) * other.getNumber(k, j);
          }
          ret.setNumber(i, j, sum);
        }
      }
      return ret;
    }
    throw new IllegalArgumentException("Unable to perform add");
  }

  /**
   * Returns the transposed matrix.
   * Rows become columns and columns become rows.
   *
   * @return transposed matrix
   */
  public Matrix transposed() {
    int[][] transposedNumbers = new int[rows][columns];
    for (int i = 0; i < rows; i++) {
      for (int j = 0; j < columns; j++) {
        transposedNumbers[j][i] = numbers[i][j];
      }
    }
    return new Matrix(transposedNumbers);
  }

  /**
   * Returns the inverse of this matrix using Gaussian elimination.
   * This method assumes the matrix is square and invertible.
   * <b>Precision loss may occur</b> due to integer arithmetic.
   *
   * @return inverted matrix
   * @throws IllegalArgumentException if the matrix is not square
   */
  public Matrix inverted() {
    if (this.rows != this.columns) {
      throw new IllegalArgumentException("Rows and columns must match");
    }
    int[][] a = numbers;
    int n = a.length;
    int[][] x = new int[n][n];
    int[][] b = new int[n][n];
    int[] index = new int[n];
    for (int i = 0; i < n; ++i) {
      b[i][i] = 1;
    }
    gaussian(a, index);
    for (int i = 0; i < n - 1; ++i) {
      for (int j = i + 1; j < n; ++j) {
        for (int k = 0; k < n; ++k) {
          b[index[j]][k] -= a[index[j]][i] * b[index[i]][k];
        }
      }
    }
    for (int i = 0; i < n; ++i) {
      x[n - 1][i] = b[index[n - 1]][i] / a[index[n - 1]][n - 1];
      for (int j = n - 2; j >= 0; --j) {
        x[j][i] = b[index[j]][i];
        for (int k = j + 1; k < n; ++k) {
          x[j][i] -= a[index[j]][k] * x[k][i];
        }
        x[j][i] /= a[index[j]][j];
      }
    }
    return new Matrix(x);
  }

  /**
   * Performs Gaussian elimination with partial pivoting.
   *
   * @param a     matrix to transform
   * @param index pivot index tracking array
   */
  private static void gaussian(int[][] a, int[] index) {
    int n = index.length;
    int[] c = new int[n];
    for (int i = 0; i < n; ++i) {
      index[i] = i;
    }
    for (int i = 0; i < n; ++i) {
      int c1 = 0;
      for (int j = 0; j < n; ++j) {
        int c0 = java.lang.Math.abs(a[i][j]);
        if (c0 > c1) {
          c1 = c0;
        }
      }
      c[i] = c1;
    }
    int k = 0;
    for (int j = 0; j < n - 1; ++j) {
      int pi1 = 0;
      for (int i = j; i < n; ++i) {
        int pi0 = java.lang.Math.abs(a[index[i]][j]) / c[index[i]];
        pi0 /= c[index[i]];
        if (pi0 > pi1) {
          pi1 = pi0;
          k = i;
        }
      }
      int temp = index[j];
      index[j] = index[k];
      index[k] = temp;
      for (int i = j + 1; i < n; ++i) {
        int pj = a[index[i]][j] / a[index[j]][j];
        a[index[i]][j] = pj;
        for (int l = j + 1; l < n; ++l) {
          a[index[i]][l] -= pj * a[index[j]][l];
        }
      }
    }
  }

  /**
   * Computes the determinant of this matrix.
   *
   * @return determinant value
   * @throws IllegalArgumentException if the matrix is not square
   */
  public double determinant() {
    if (rows != columns) {
      throw new IllegalArgumentException("Matrix must be square to compute determinant");
    }
    if (rows == 1) {
      return numbers[0][0];
    } else if (rows == 2) {
      return numbers[0][0] * numbers[1][1] - numbers[0][1] * numbers[1][0];
    } else if (rows == 3) {
      int x = numbers[0][0] * (numbers[1][1] * numbers[2][2] - numbers[1][2] * numbers[2][1]);
      int y = numbers[0][1] * (numbers[1][0] * numbers[2][2] - numbers[1][2] * numbers[2][0]);
      int z = numbers[0][2] * (numbers[1][0] * numbers[2][1] - numbers[1][1] * numbers[2][0]);
      return x - y + z;
    } else {
      double[][] temp = new double[rows][columns];
      for (int i = 0; i < rows; i++) {
        for (int j = 0; j < columns; j++) {
          temp[i][j] = numbers[i][j];
        }
      }
      double det = 1.0;

      for (int i = 0; i < rows; i++) {
        int pivot = i;
        for (int j = i + 1; j < rows; j++) {
          if (java.lang.Math.abs(temp[j][i]) > java.lang.Math.abs(temp[pivot][i])) {
            pivot = j;
          }
        }
        if (temp[pivot][i] == 0) {
          return 0;
        }
        if (i != pivot) {
          double[] swap = temp[i];
          temp[i] = temp[pivot];
          temp[pivot] = swap;
          det *= -1;
        }
        double pivotVal = temp[i][i];
        det *= pivotVal;
        for (int j = i + 1; j < rows; j++) {
          double factor = temp[j][i] / pivotVal;
          for (int k = i; k < columns; k++) {
            temp[j][k] -= factor * temp[i][k];
          }
        }
      }
      return det;
    }
  }

  public void print() {
    for (int i = 0; i < rows; i++) {
      for (int j = 0; j < columns; j++) {
        System.out.print(numbers[i][j] + " ");
      }
      System.out.println();
    }
  }
}

File: src/main/java/org/example/runtime/extra/matrix.cpp
#include "matrix.hpp"
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "../allocator/allocator.hpp"
#include "../error/error.hpp"

Matrix matrix_init(int rows, int columns) {
    Matrix matrix;
    matrix.rows = rows;
    matrix.colums = columns;
    matrix.data = (int**)allocateFromMarsik(rows * sizeof(int*));
    for (int i = 0; i < rows; i++) {
        matrix.data[i] = (int*)allocateFromMarsik(columns * sizeof(int));
        for (int j = 0; j < columns; j++) {
            matrix.data[i][j] = 0;
        }
    }
    return matrix;
}

int matrix_get(Matrix* matrix, int row, int column) {
    if (row < 0 || row > matrix->rows) {
        runtimeError("Row does not match matrix");
        return -1;
    }
    if (column < 0 || column > matrix->colums ) {
        runtimeError("Column does not match matrix");
        return -1;
    }
    return matrix->data[row][column];
}

void matrix_set(Matrix* matrix, int row, int column, int value) {
    if (row < 0 || row > matrix->rows) {
        runtimeError("Row does not match matrix");
        return;
    }
    if (column < 0 || column > matrix->colums ) {
        runtimeError("Column does not match matrix");
        return;
    }
    matrix->data[row][column] = value;
}

int matrix_numberOfRows(Matrix* matrix) {
    return matrix->rows;
}

int matrix_numberOfColumns(Matrix* matrix) {
    return matrix->colums;
}

Matrix matrix_add(Matrix* matrix, Matrix* other) {
    if (matrix->rows == other->rows && matrix->colums == other->colums) {
        Matrix newMatrix;
        newMatrix.data = (int**)allocateFromMarsik(matrix->rows * sizeof(int*));
        newMatrix.rows = matrix->rows;
        newMatrix.colums = matrix->colums;
        for (int i = 0; i < newMatrix.rows; i++) {
            newMatrix.data[i] = (int*)allocateFromMarsik(matrix->colums * sizeof(int));
            for (int j = 0; j < newMatrix.colums; j++) {
               matrix_set(&newMatrix, i, j, matrix_get(matrix, i, j) + matrix_get(other, i, j));
        }
      }
      return newMatrix;
    } else {
        runtimeError("Matrix dimensions does not match for addition");
        return matrix_init(matrix->rows, matrix->colums);
    }
}

Matrix matrix_multiply(Matrix* matrix, Matrix* other) {
    if (matrix->rows == other->colums && matrix->colums == other->rows) {
        Matrix newMatrix;
        newMatrix.data = (int**)allocateFromMarsik(matrix->rows * sizeof(int*));
        newMatrix.rows = matrix->rows;
        newMatrix.colums = matrix->colums;
        for (int i = 0; i < newMatrix.rows; i++) {
            newMatrix.data[i] = (int*)allocateFromMarsik(matrix->colums * sizeof(int));
            for (int j = 0; j < newMatrix.colums; j++) {
               int sum = 0;
                for (int k = 0; k < matrix->colums; k++) {
                 sum += matrix_get(matrix, i, k) * matrix_get(other, k, j);
                }
            matrix_set(&newMatrix, i, j, sum);    
        }
      }
      return newMatrix;
    } else {
        runtimeError("Matrix dimensions does not match for multiplication");
        return matrix_init(matrix->rows, matrix->colums);
    }
}

Matrix matrix_transpose(Matrix* matrix) {
    Matrix transposed = matrix_init(matrix->rows, matrix->colums);
    for (int i = 0; i < matrix->rows; i++) {
      for (int j = 0; j < matrix->colums; j++) {
        transposed.data[j][i] = matrix->data[i][j];
      }
    }
    return transposed;
}

Matrix matrix_clone(Matrix* matrix) {
    Matrix cloned = matrix_init(matrix->rows, matrix->colums);
    for (int i = 0; i < matrix->rows; i++) {
      for (int j = 0; j < matrix->colums; j++) {
        cloned.data[i][j] = matrix->data[i][j];
      }
    }
    return cloned;
}

double matrix_getDeterminant(Matrix* matrix) {
    if (matrix->colums != matrix->rows) {
        runtimeError("Matrix must be square");
        return 0;
    }
    if (matrix->rows == 1) {
        return matrix_get(matrix, 0, 0);
    }
    if (matrix->rows == 2) {
        return matrix_get(matrix, 0, 0) * matrix_get(matrix, 1, 1) -
               matrix_get(matrix, 0, 1) * matrix_get(matrix, 1, 0);
    }
    if (matrix->rows == 3) {
        int x = matrix_get(matrix, 0, 0) * (matrix_get(matrix, 1, 1) * matrix_get(matrix, 2, 2) 
            - matrix_get(matrix, 1, 2) * matrix_get(matrix, 2, 1));
        int y = matrix_get(matrix, 0, 1) * (matrix_get(matrix, 1, 0) * matrix_get(matrix, 2, 2)
            - matrix_get(matrix, 1, 2) * matrix_get(matrix, 2, 0));
        int z = matrix_get(matrix, 0, 2) * (matrix_get(matrix, 1, 0)* matrix_get(matrix, 2, 1) 
            - matrix_get(matrix, 1, 1) * matrix_get(matrix, 2, 0));
        return x - y + z;
    }
    else {
        int rows = matrix->rows;
        int columns = matrix->colums;
        double temp[rows][columns];
        for (int i = 0; i < rows; i++) {
          for (int j = 0; j < columns; j++) {
             temp[i][j] = matrix_get(matrix, i, j);
          }
        }
        double det = 1.0;

        for (int i = 0; i < rows; i++) {
          int pivot = i;
          for (int j = i + 1; j < rows; j++) {
            if (fabs(temp[j][i]) > fabs(temp[pivot][i])) {
              pivot = j;
          }
        }
        if (temp[pivot][i] == 0) {
          return 0;
        }
        if (i != pivot) {
          for (int j = 0; j < 10; j++) {
            double swap = temp[i][j];
            temp[i][j] = temp[pivot][j];
            temp[pivot][j] = swap;
          }
        }
        double pivotVal = temp[i][i];
        det *= pivotVal;
        for (int j = i + 1; j < rows; j++) {
          double factor = temp[j][i] / pivotVal;
          for (int k = i; k < columns; k++) {
            temp[j][k] -= factor * temp[i][k];
          }
        }
      }
      return det;
    } 
}

void matrix_print(Matrix* matrix) {
    for (int i = 0; i < matrix->rows; i++) {
      for (int j = 0; j < matrix->colums; j++) {
        printf("%d", matrix->data[i][j]);
      }
      printf(" \n");
    }
}